### ISSUE TITLE
False-positive for rule `no-this-in-sfc`

### ISSUE THREAD

**Issue by markogresak** [yannickcr/eslint-plugin-react#2147]
This code produces a false-positive error:
``` jsx
import React, { Fragment } from 'react';

const CustomComponent = () => {};

const Example = ({ prop }) => {
    return {
        handleClick: () => {
            console.log('click');
        },

        renderNode({ children, value }) {
            return (
                <Fragment>
                    {children}
                    <CustomComponent 
                        onClick={this.handleClick} 
                        value={value} 
                        prop={prop} 
                    />
                </Fragment>
            );
        },
    };
};

export default Example;
```
```
example.jsx:15:47: Stateless functional components should not use this [Error/react/no-this-in-sfc]
```

Problems:
1. The `Example` is not an SFC. It returns a plain object.
2. The `renderNode` is a render function, the `this.handleClick` is valid reference to the `handleClick` function on the object returned by `Example`.

Expected behaviour:
The rule correctly detects the context of the render function and doesn't treat it as an SFC. 

A smaller, console-testable example to show that the above code works:
``` js
const X = {
    getRandomValue: () => 4,
    double() {
        return this.getRandomValue() * 2;
    },
};
console.log(X.double());
```

Using:
 - `eslint-plugin-react@7.12.4`
 - `eslint@5.12.1`
 - rule config: `"react/no-this-in-sfc": "error"`

**Comment by ljharb** [comment]
Typically, render functions should be SFCs - however I agree that in this case (an object literal member, and a name that doesn't start with a capital letter) it shouldn't be detected as a component.

**Comment by mjdavidson** [comment]
false positive for https://github.com/sequelize/sequelize models, need to use 
```javascript 
this.getDataValue('deadline')
```
to get the underlying data value
